text/0000-delegation.md

```diff
@@ -0,0 +1,612 @@
+```rust
+impl TR for S {
+    type Item = <F as TR>::Item;
+    // ...
+    const MAX = <F as TR>::MAX;
+    // ...
+    fn foo(&self) -> u32 {
+        self.f.foo()
+    }
+    fn bar(&self, x: u32, y: u32, z: u32) -> u32 {
+        self.f.bar(x, y, z)
+    }
+    // ...
+}
+```
+
+To delegate most of a trait, rather than all of it, simply `delegate *` and then write the manual implementations for the items you don’t want to delegate.
+
+```rust
+impl TR for S {
+    delegate * to f;
+
+    fn foo(&self) -> u32 {
+        42
+    }
+}
+
+```
+
+Aside from the implementation of foo(), this has exactly the same meaning as the first example.
+
+If you only want to delegate specific items, rather than “all” or “most” items, then replace `*` with a comma-separated list of only the items you want to delegate. Since it’s possible for types and functions to have the same name, the items must be prefixed with `fn`, `const` and `type` as appropriate.
+
+```rust
+impl TR for S {
+    delegate fn foo, fn bar, const MAX, type Item 
+        to f;
+}
+```
+
+This also has the exact same meaning as the first example. While `to field_name;` is not required to be on a new line, this is the `rustfmt` default when delegating individual trait items, as it makes it easy to visually identify the field being delegated to.
+
+If you want to delegate some methods to one field and some to another, simply write multiple `delegate` statements:
+
+```rust
+impl TR for S {
+    delegate * to field_one;
+    delegate fn foo, const MAX, type Item
+        to field_two;
+}
+```
+
+This delegates `foo`, `MAX`, and `Item` to `field_two`. Everything else is delegated to `field_one`.
+
+In addition to "regular" structs with named fields, you can delegate to indexed fields on a tuple struct. This first field name is `0`, which we're used to writing as `self.0`:
+
+```rust
+struct AddOnlyCounter(u32);
+
+impl AddOnlyCounter {
+    fn new() -> Self {
+        Self { 0 }
+    }
+}
+
+impl AddAssign for AddOnlyCounter {
+    delegate * to 0;
+}
+```
+
+By using the _newtype pattern_, we _restrict_ the behaviour of the original type and only allow semantically valid methods/operations. If `AddOnlyCounter` is defined in another module, users can't break its guarantees, as the `u32` is private aka encapsulated.
+
+```rust
+let counter = AddOnlyCounter::new();
+
+// this works:
+counter += 1;
+
+// these are compile-time errors:
+counter = 5;
+counter -= 1;
+counter.0 -= 1;
+```
+
+
+# Reference-level explanation
+[reference]: #reference
+
+A delegation statement can only appear inside a trait impl block. Delegation inside inherent impls is left as a future extension.
+
+Delegation must be to a field on `Self`. Other kinds of implementer expressions are left as future extensions. This also means delegation can only be done on structs for now.
+
+Delegation statements must be the first items in an impl block. There may be more than one delegation statement, but they must all be at the top.
```

Comment:
Based on my memory of past and current delegation threads:

1) Some people thought this was required to make the contextual keyword option viable. I think this line of reasoning predates the Great Epoch Debate. It's obviously obsolete now that we have official guidance on new keyword introduction.

2) When an `impl` block contains a delegation as well as a regular item implementation, it seems better style to put the delegation on top. Especially when it's a `*` delegation, since "Foo implements trait T by delegating almost everything to field f, except that method foo is ... " is the natural way to describe that sort of thing.

So imo, as part of editing this to say `delegate` should be a real keyword, we should also delete this restriction. Putting the `delegate` item(s) at the top is probably still good style, but there's no need to make it mandatory.